To: Mike Frysinger <vapier@gentoo.org>
Cc: bug-coreutils@gnu.org
Subject: Re: regression in `ls -x` introduced between 6.7 and 6.8
From: Paul Eggert <eggert@CS.UCLA.EDU>
Date: Sun, 01 Apr 2007 20:31:14 -0700

Mike Frysinger <vapier@gentoo.org> writes:

>> Do you observe a similar problem with ls -m?
>
> nope

OK, thanks, that helped me to narrow it down.  Here is a patch.

2007-04-01  Paul Eggert  <eggert@cs.ucla.edu>

	* src/ls.c (print_horizontal): Fix bug reported by Mike Frysinger:
	ls -x DIR would sometimes output the wrong string in place of the
	first entry.
	* tests/ls/x-option: New file.
	* tests/ls/Makefile.am (TESTS): Add x-option.

diff --git a/src/ls.c b/src/ls.c
index b33146c..abb0a26 100644
--- a/src/ls.c
+++ b/src/ls.c
@@ -4074,16 +4074,16 @@ print_horizontal (void)
   size_t pos = 0;
   size_t cols = calculate_columns (false);
   struct column_info const *line_fmt = &column_info[cols - 1];
-  size_t name_length = length_of_file_name_and_frills (cwd_file);
+  struct fileinfo const *f = sorted_file[0];
+  size_t name_length = length_of_file_name_and_frills (f);
   size_t max_name_length = line_fmt->col_arr[0];

   /* Print first entry.  */
-  print_file_name_and_frills (cwd_file);
+  print_file_name_and_frills (f);

   /* Now the rest.  */
   for (filesno = 1; filesno < cwd_n_used; ++filesno)
     {
-      struct fileinfo const *f;
       size_t col = filesno % cols;

       if (col == 0)
diff --git a/tests/ls/Makefile.am b/tests/ls/Makefile.am
index 461ba07..2b6fb90 100644
--- a/tests/ls/Makefile.am
+++ b/tests/ls/Makefile.am
@@ -24,7 +24,7 @@ TESTS = \
   stat-dtype \
   inode dangle file-type recursive dired infloop \
   rt-1 time-1 symlink-slash follow-slink no-arg m-option \
-  stat-vs-dirent
+  stat-vs-dirent x-option

 EXTRA_DIST = $(TESTS)
 TESTS_ENVIRONMENT = \
diff --git a/tests/ls/x-option b/tests/ls/x-option
new file mode 100755
index 0000000..44556df
--- /dev/null
+++ b/tests/ls/x-option
@@ -0,0 +1,59 @@
+#!/bin/sh
+# Exercise the -x option.
+
+# Copyright (C) 2007 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.
+
+if test "$VERBOSE" = yes; then
+  set -x
+  ls --version
+fi
+
+. $srcdir/../envvar-check
+. $srcdir/../lang-default
+
+pwd=`pwd`
+t0=`echo "$0"|sed 's,.*/,,'`.tmp; tmp=$t0/$$
+trap 'status=$?; cd "$pwd" && chmod -R u+rwx $t0 && rm -rf $t0 && exit $status' 0
+trap '(exit $?); exit $?' 1 2 13 15
+
+framework_failure=0
+mkdir -p $tmp || framework_failure=1
+cd $tmp || framework_failure=1
+mkdir subdir || framework_failure=1
+touch subdir/b || framework_failure=1
+touch subdir/a || framework_failure=1
+
+if test $framework_failure = 1; then
+  echo "$0: failure in testing framework" 1>&2
+  (exit 1); exit 1
+fi
+
+fail=0
+
+# Coreutils 6.8 and 6.9 would output this in the wrong order.
+ls -x subdir > out || fail=1
+ls -rx subdir >> out || fail=1
+cat <<\EOF > exp || fail=1
+a  b
+b  a
+EOF
+
+cmp out exp || fail=1
+test $fail = 1 && diff out exp 2> /dev/null
+
+(exit $fail); exit $fail
M ChangeLog
M src/ls.c
M tests/ls/Makefile.am
A tests/ls/x-option
Committed as dd35c3cab3ada1120a14b1292b0ac5c60357afb3

